synphot/tests/test_observation.py

```diff
@@ -439,3 +440,136 @@ def test_countrate_neg_leak():
+    answer = 342
+
+    # allow error to be +/- 1 for this test
+    assert_quantity_allclose(result, answer, atol=1)
+
+
+def test_snr_bright_object():
+    """
+    Test that observation.ccd_snr returns sqrt(counts), the expected value
+    for a bright target.
+    """
+    counts = 25e5 * u.ct
+    result = ccd_snr(counts)
+    answer = np.sqrt(counts.value)
+
+    assert_quantity_allclose(result, answer)
+
+
+def test_t_exp_numeric():
+    """
+    A test to check that the numerical method in
+    observation.exptime_from_ccd_snr (i.e. when the error in background
+    noise or gain is non-negligible) is done correctly. Based on the worked
+    example on pg. 56-57 of [1]_.
+
+    References
+    ----------
+    .. [1] Howell, S. B. 2000, *Handbook of CCD Astronomy* (Cambridge, UK:
+        Cambridge University Press)
+    """
+    t = 300 * u.s
+    snr = 342
+    gain = 5 * (u.ct / u.adu)
+    countrate = 24013 * u.adu * gain / t
+    npix = 1 * u.pixel
+    n_background = 200 * u.pixel
+    background_rate = 620 * (u.adu / u.pixel) * gain / t
+    darkcurrent_rate = 22 * (u.ct / u.pixel / u.hr)
+    readnoise = 5 * (u.ct / u.pixel)
+
+    result = exptime_from_ccd_snr(snr, countrate, npix=npix,
+                                  n_background=n_background,
+                                  background_rate=background_rate,
+                                  darkcurrent_rate=darkcurrent_rate,
+                                  readnoise=readnoise, gain=gain)
+
+    assert_quantity_allclose(result, t, atol=1 * u.s)
+
+
+def test_t_exp_analytic():
+    """
+    A test to check that the analytic method in
+    observation.exptime_from_ccd_snr is done correctly.
+    """
+    snr_set = 50
+    countrate = 1000 * (u.ct / u.s)
+    npix = 1 * u.pixel
+    background_rate = 100 * (u.ct / u.pixel / u.s)
+    darkcurrent_rate = 5 * (u.ct / u.pixel / u.s)
+    readnoise = 1 * (u.ct / u.pixel)
+
+    t = exptime_from_ccd_snr(snr_set, countrate, npix=npix,
+                             background_rate=background_rate,
+                             darkcurrent_rate=darkcurrent_rate,
+                             readnoise=readnoise)
+
+    # if t is correct, ccd_snr() should return snr_set:
+    snr_calc = ccd_snr(countrate * t,
+                       npix=npix,
+                       background=background_rate * t,
+                       darkcurrent=darkcurrent_rate * t,
+                       readnoise=readnoise)
+
+    assert_quantity_allclose(snr_calc, snr_set,
+                             atol=0.5 * u.Unit(""))
+
+
+def test_snr_with_countrate():
+    """
+    A test to make sure `ccd_snr` works with the units that
+    `countrate` returns.
+    """
+    source_spec = SourceSpectrum(BlackBodyNorm1D)
+    bp = SpectralElement(Const1D)
+    obs = Observation(source_spec, bp, force='extrap')
+    counts = obs.countrate(1 * u.m ** 2) * 1 * u.s
+    snr = ccd_snr(counts)
+
+    assert snr.unit == u.Unit("")
```

Comment:
I think SNR should be close to 1, as `1/sqrt(1)` is 1, but I'll leave that as an exercise for you.
```suggestion
    assert snr.unit == u.Unit("")
    np.testing.assert_allclose(snr.value, ???)
```